aws-android-sdk-cognitoauth/src/main/java/com/amazonaws/mobileconnectors/cognitoauth/AuthClient.java

```diff
@@ -709,21 +713,68 @@ private void launchSignOut(final String redirectUri, final String browserPackage
         launchCustomTabs(fqdn, null, browserPackage);
     }
 
+    /***
+     * Check if a browser is installed on the device to launch HostedUI.
+     * @return true if a browser exists else false.
+     */
+    public Boolean isBrowserInstalled() {
+        String url = "https://docs.amplify.aws/";
+        Uri webAddress = Uri.parse(url);
+        Intent intentWeb = new Intent(Intent.ACTION_VIEW, webAddress);
+        return (intentWeb.resolveActivity(context.getPackageManager()) != null);
+    }
+
+    /**
+     * Get list of browser packages that support Custom Tabs Service.
+     * @return list of package names that support Custom Tabs.
+     */
+    private ArrayList<String> getSupportedBrowserPackage(){
+        PackageManager packageManager = context.getPackageManager();
+        // Get default VIEW intent handler.
+        Intent activityIntent = new Intent()
+                .setAction(Intent.ACTION_VIEW)
+                .addCategory(Intent.CATEGORY_BROWSABLE)
+                .setData(Uri.fromParts("http", "", null));
+
+        // Get all apps that can handle VIEW intents.
+        List<ResolveInfo> resolvedActivityList = packageManager.queryIntentActivities(activityIntent, 0);
+        ArrayList<String> packageNamesSupportingCustomTabs = new ArrayList<>();
+        for (ResolveInfo info : resolvedActivityList) {
+            Intent serviceIntent = new Intent()
+                    .setAction(ACTION_CUSTOM_TABS_CONNECTION)
+                    .setPackage(info.activityInfo.packageName);
+            // Check if this package also resolves the Custom Tabs service.
+            if(packageManager.resolveService(serviceIntent, 0) != null) {
+                packageNamesSupportingCustomTabs.add(info.activityInfo.packageName);
+            }
+        }
+        return packageNamesSupportingCustomTabs;
+    }
+
     /**
      * Launches the HostedUI webpage on a Custom Tab.
      * @param uri Required: {@link Uri}.
      * @param activity Activity to launch custom tabs from and which will listen for the intent completion.
      * @param browserPackage Optional string specifying the browser package to launch the specified url.
-     *                       Defaults to Chrome if null.
+     *                       Launches intent chooser if set to null.
      */
     private void launchCustomTabs(final Uri uri, final Activity activity, final String browserPackage) {
     	try {
+    	    if(!isBrowserInstalled()) {
+                userHandler.onFailure(new Exception("No browsers installed."));
```

Comment:
We should use `RuntimeException` to be (slightly) more specific. Same for when custom tabs browser isn't found. It is generally best practice to use as specific of an exception type as possible (even if it isn't followed throughout this project unfortunately).